EIPS/eip-6865.md

```diff
@@ -0,0 +1,151 @@
+- `amounts` is an Array of `uint256` where items MUST define the amount per time curve, with time defined within liveness boundaries
+    - the first amount in `amounts` Array (amounts[0]) MUST be the amount of the asset at `liveness.from` timestamp
+    - the last amount in `amounts` Array (amounts[amounts.length-1]) MUST be the amount of the asset at `liveness.to` timestamp
+    - in most of the cases, `amounts` will be an Array with a single item which is MUST be the minimum amount of the asset.
+
+#### `assetsIn`
+
+`assetsIn` are the minimum assets which the user MUST get if the message is signed and fulfilled
+
+#### `assetsOut`
+
+`assetsOut` are the maximum assets which the user MUST offer if the message is signed and fulfilled
+
+## Rationale
+
+### on-chain
+
+One might argue that certain processes can be done off-chain, which is true, but our experience building an off-chain TypeScript SDK to solve this matter revealed some issues:
+
+- Reliability: Protocols developers are the ones responsible for developing the protocol itself, thus crafting the visualization is much more accurate when done by them.
+- Scalability: Keeping up with the rapidly expanding ecosystem is hard. Wallets or 3rd party entities must keep an eye on each new protocol, understand it carefully (which poses the reliability issues mentioned above), and then only come up with an off-chain implementation.
+- Maintainability: Many protocols implement smart contracts in an upgradable manner. This causes the off-chain visualization to differ from the real protocol behaviors (if updated), making the solution itself unreliable and lacking the scalability to handle various protocols.
+
+### `DomainHash`
+
+The `domainHash` is much needed by protocols to revert against unsupported versions of its EIP-712 implementation. It identifies the needed implementation in case the protocol implements various EIP-712 implementations (`name`) or to revert if the `domainHash` belongs to a different protocol.
+
+In the future, if there is a registry that reroutes this EIP implementation for already deployed protocols that can't upgrade the existing deployed smart contract, `domainHash` can be used to identify protocols.
+
+### Amounts Array
+
+We suggest using an array of amounts (uint256[]) instead of a single uint256 to cover auctions, which are common in NFT protocols.
+
+## Backwards Compatibility
+
+No backward compatibility issues found.
+
+## Reference Implementation
+
+openSea Seaport NFT marketplace implementation example is available [here](../assets/eip-6865/contracts/SeaPortEIP712Visualizer.sol)
+
+## Security Considerations
+
+`visualizeEIP712Message` function should be reliable and accurately represent the potential impact of the EIP-712 message on users' assets. Wallet providers and users must trust the protocol's implementation of this function to provide accurate and up-to-date information.
```

Comment:
Normally I wouldn't hold up a proposal because of a security consideration, but I think this needs to be addressed even before going into draft (because proposals in draft sometimes stay in draft even after implementations are released...)

What happens if I'm transacting with a malicious website `obviousscam.invalid`, and `obviousscam.invalid` pops up a sign typed data request. What in this proposal stops `obviousscam.invalid`'s contract from showing a completely different visualization from what the real contract will do?

If it's the `verifyingContract` in the domain separator, you should mention that signing requests without that field do not get visualized.

Reply:
also, users usually and wallets SHOULD always highlight to user the nature and reputation of the smart contract they are interacting with, let's take an example to make it clear:
user head up to `obviousscam.invalid` where this scammy website is trying to steal the funds, and i can imagine two possible scenarios using EIP-712:
## 1
`obviousscam.invalid` ask user to sign an EIP-712 where the `verifyingContract` is `0xbad` which is the scam website smart contract itself, at this stage user didn't give any approval to `0xbad` or any address associated with this entity, because if s/he did, there's no need even to reach this point because the scammy entity would have already rekt the user.
means even if the `0xbad` lies about it's behaviour to scam user, there's no real impact at that time because the EIP-712 message is intended to that entity itself only, and cant be used elsewhere.
indeed, to even cut it at that level, wallets usually(and SHOULD always) show the verified sign when it comes to smart contract addresses, so a known protocol smart contract like openSea, looksrare, blur, curve..etc are displayed as verified/trusted smart contract (code is low), but less-knows to unknown protocols/smart contracts are always displayed as raw addresses and users usually (should be) warned that this smart contract is unknown and s/he must double check what s/he is doing (this applies to all crypto space, even RPC simulation can be misleading). 

##2 
`obviousscam.invalid` ask user to sign an EIP-712 where the `verifyingContract` is some other protocol, for example opensea marketplace (which is the case of many phishing attacks , you can check [this $1.07 million attack](https://twitter.com/Serpent/status/1604074440941506560?lang=en)
in this case wallets uses `verifyingContract` which points to the real opensea protocol address, they show that this EIP-712 message is intended for opensea (which is already suspicious to user), wallet call [this proposal](https://github.com/ethereum/EIPs/pull/6865) smart contact implementation and gets the `Result` that consist of `assetsIn` and `assetsOut` which can be easily displayed/analysed on the front-side 

TLDR; `verifyingContract` should be displayed based on the address reputation which is something already in place for every smart contract (on-chain transactions), and the EIP-712 message should be treated based on the reputation of its `verifyingContract`, any EIP-712 message that doesn't have a `verifyingContract` is ignored and honestly i don't see a case of an EIP-712 message that needs to be executed on-chain without the `verifyingContract`

Reply:
will update the PR accordingly and request you review again, Thanks!

Reply:
add more details [here](https://github.com/ethereum/EIPs/pull/6865/commits/d712f865c9cd9ff18144cd1c72d94c8068799043)